_docs/implementation.md

```diff
@@ -0,0 +1,183 @@
+in Go programs, enabling dynamic monitoring and modification of function behavior.
+The approach leverages trampoline code injection and function pointer redirection
+to seamlessly integrate monitoring logic without requiring significant changes to
+the user's codebase.
+
+The goal is to provide a flexible and non-intrusive mechanism for instrumenting
+Go applications, particularly for use cases such as observability (e.g., OpenTelemetry),
+debugging, security or performance profiling.
+
+# Core Principles
+
+## 1. Trampoline Code Injection
+
+Trampoline function calls are injected into the Target (lib-side) function, which
+ultimately jumps to the actual hook code via the function pointer Hook. The two-level
+indirection (trampoline call to hook) is a deliberate design choice that offers
+several key benefits:
+
+- Exception Handling: The trampoline catches panics and isolates exception handling,
+  preventing them from affecting the target function or hook code.
+- Context Construction: The trampoline initializes and manages the necessary
+  context before invoking the hook code.
+- Decoupling: The trampoline decouples the hook code from the target function,
+  enabling flexibility and maintainability.
+- Debugging: The trampoline provides a centralized point for debugging and observability.
+- Dynamic Instrumentation: The trampoline allows turn-on/off instrumentation
+  at runtime, enabling dynamic control over monitoring behavior.
+
+The code snippet below illustrates the two-level indirection, where the Target
+function is the target function to be instrumented, and the Hook function is the
+monitoring code that will be executed.
+
+```go
+func Target() {
+    Trampoline()
+    ....
+}
+
+func Trampoline() {
+    Hook()
+}
+
+//go:linkname Hook github.com/open-telemetry/opentelemetry-go-compile-instrumentation/sdk/hook.MyHook
+var Hook func()
+```
+
+
+## 2. Linkage via golinkname
+
+The `Hook` function is linked to the monitoring code using the `//go:linkname`
+directive. This allows us to dynamically associate the hook with the target
+function at compile time.
+
+# Implementation Details
+
+Go compile-time instrumentation is a two-phase process: the first phase involves
+setup dependencies, and the second phase focuses on building the project using
+a custom toolchain.
+
+## Phase 1: Setup Dependencies
+
+### 1.1 Dependency Analysis
+The first step is to analyze the project's dependencies by collecting the list 
+of modules involved in the build. This is done using the `go build -n` command, 
+which prints the build plan without executing it.
+
+```command
+$ go build -n > build_plan.txt
+```
+
+From the build_plan.txt file, the tool extracts all third-party module paths 
+(e.g., `github.com/go-redis/redis`, `github.com/gin-gonic/gin`).
+
+## 1.2 Add Dependencies
+Once the third-party dependencies are identified, the tool generates a file 
+(e.g., otel_import.go) to import the SDK and corresponding hook packages for each 
+dependency:
```

Comment:
I think this forgets to explain how available hook packages/instrumentation opportunities are identified... Doesn't need to be too detailed, but I am curious what you have in mind.